core-v1/src/main/scala/engine/SimpleCachingEngine.scala

```diff
@@ -30,8 +30,20 @@ object SimpleCachingEngine {
   )
 
   final object CachedResult {
-    implicit val instance: Extract[CachedResult] with Foldable[CachedResult] with Functor[CachedResult] =
-      new Extract[CachedResult] with Foldable[CachedResult] with Functor[CachedResult] {
+    implicit val instance: Applicative[CachedResult]
+      with Extract[CachedResult]
+      with Foldable[CachedResult]
+      with Functor[CachedResult] =
+      new Applicative[CachedResult] with Extract[CachedResult] with Foldable[CachedResult] with Functor[CachedResult] {
+
+        override def pure[A](x: A): CachedResult[A] = CachedResult(x, Map.empty)
+
+        override def ap[A, B](ff: CachedResult[A => B])(fa: CachedResult[A]): CachedResult[B] = {
+          val value = ff.value(fa.value)
+          // TODO: What if I just take the cache from fa?
+          val combinedCache = ff.cacheState ++ fa.cacheState
+          CachedResult(value, combinedCache)
+        }
```

Comment:
I think accepting the cache from only the final result would work in almost every case.

It is possible that the implementation of the given `ff` is incorrect, in that it does not pass the `cacheState` from `ff` into the result of `fa`. This would result in the final result missing some state from the previous cache, but I think the definition of `Applicative` would not be violated in this case.

Only accepting the cache state from `fa` would violate the definition of `Parallel`, as the order of monadic chaining matters for determining what the final cache state is, however, this class is useless in a parallel manner. The whole point of a cache is to remember the result of previous actions, so the order of actions is always relevant and the cache must always be updated between operations. Executing expressions in parallel means that we cannot take advantage of this type of cache. It would be possible to share a cache of atomic pointers in the future, but we would have to reimplement this interpreter and model class anyway.

I think I'm going to just return `fa.cacheState`